# NumPy: Tutorial Completo

**Objetivo:** Dominar NumPy desde fundamentos hasta técnicas avanzadas para Machine Learning y Data Science.

---

## Tabla de Contenidos

1. Introducción a NumPy
2. Creación de Arrays
3. Atributos y Propiedades de Arrays
4. Indexación y Slicing
5. Operaciones Matemáticas y Broadcasting
6. Funciones Universales (ufuncs)
7. Manipulación de Forma y Dimensiones
8. Operaciones Estadísticas
9. Álgebra Lineal
10. Números Aleatorios
11. Lectura y Escritura de Datos
12. Técnicas Avanzadas
13. Ejercicios Prácticos

---

## 1. Introducción a NumPy

NumPy (Numerical Python) es la librería fundamental para computación científica en Python. Proporciona:

- Arrays multidimensionales eficientes (ndarray)
- Operaciones vectorizadas de alto rendimiento
- Funciones matemáticas, estadísticas y de álgebra lineal
- Base para librerías como pandas, scikit-learn, TensorFlow

### ¿Por qué NumPy?

- **Velocidad:** Implementado en C, 10-100x más rápido que listas Python
- **Memoria:** Uso eficiente de memoria
- **Sintaxis:** Operaciones vectorizadas sin loops explícitos

In [1]:
import numpy as np

# Verificar versión
print(f"NumPy version: {np.__version__}")

NumPy version: 2.4.2


### Comparación: Lista vs NumPy Array

In [8]:
import time

# Lista Python
lista = list(range(1000000))
start = time.time()
lista_cuadrados = [x**20 for x in lista]
tiempo_lista = time.time() - start

# NumPy Array
array = np.arange(1000000)
start = time.time()
array_cuadrados = array**20
tiempo_numpy = time.time() - start

print(f"Tiempo con listas: {tiempo_lista:.4f}s")
print(f"Tiempo con NumPy: {tiempo_numpy:.4f}s")
print(f"NumPy es {tiempo_lista/tiempo_numpy:.1f}x más rápido")

Tiempo con listas: 0.3128s
Tiempo con NumPy: 0.0030s
NumPy es 105.0x más rápido


---

## 2. Creación de Arrays

### 2.1 Desde listas y tuplas

In [2]:
# Array 1D desde lista

list1d = [1, 4, 1, 2]
arr1d = np.array(list1d)
print("Array 1D:", arr1d)

arr2d = np.array([1, 2, 3, 4, 5])
print("Array 2D:", arr2d)

# Array 1D desde tupla
arr1d_tuple = np.array((1, 2, 3, 4, 5))
print("\nArray 1D desde tupla:", arr1d_tuple)

# Array 2D desde lista de listas (matriz)
arr2d = np.array([[1, 2, 3], [4, 5, 6]])
print("\nArray 2D:")
print(arr2d)

# Array 3D
arr3d = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
print("\nArray 3D:")
print(arr3d)

Array 1D: [1 4 1 2]
Array 2D: [1 2 3 4 5]

Array 1D desde tupla: [1 2 3 4 5]

Array 2D:
[[1 2 3]
 [4 5 6]]

Array 3D:
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


In [3]:
class Matrix:
    def __init__(self, rows, cols):
        self.rows = rows
        self.cols = cols
    
    def create_matrix(self) -> np.ndarray:
        return np.zeros((self.rows, self.cols))
    
matrix = Matrix(3, 4)
print("\nMatriz creada con clase Matrix:")
print(matrix.create_matrix())


Matriz creada con clase Matrix:
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


### 2.2 Funciones de creación

In [18]:
# Ceros
zeros = np.zeros((3, 4)) # rows=3, cols=4
print("Zeros (3x4):")
print(zeros)

# Unos
ones = np.ones((2, 3, 4)) # 2 matrices de 3x4 llenas de unos
print("\nOnes (2x3x4):\n", ones)

# Valor constante
full = np.full((3, 3), 7) # matriz de 3x3 llena de 7s
print("\nFull de 7s (3x3):")
print(full)

# Matriz identidad
identity = np.eye(4)
print("\nMatriz identidad (4x4):")
print(identity)

i = 0
f = 50
p = 5
# Rango de valores
arange_arr = np.arange(i, f, p)  # inicio, fin, paso
print(f"\nArange ({i} a {f}, paso {p}):", arange_arr)

i = 0
f = 50
c = 5
# Espacio lineal
linspace_arr = np.linspace(i, f, c)  # inicio, fin, cantidad de puntos
print(f"\nLinspace ({i} a {f}, {c} puntos):", linspace_arr)

Zeros (3x4):
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]

Ones (2x3x4):
 [[[1. 1. 1. 1.]
  [1. 1. 1. 1.]
  [1. 1. 1. 1.]]

 [[1. 1. 1. 1.]
  [1. 1. 1. 1.]
  [1. 1. 1. 1.]]]

Full de 7s (3x3):
[[7 7 7]
 [7 7 7]
 [7 7 7]]

Matriz identidad (4x4):
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]

Arange (0 a 50, paso 5): [ 0  5 10 15 20 25 30 35 40 45]

Linspace (0 a 50, 5 puntos): [ 0.  12.5 25.  37.5 50. ]


### 2.3 Arrays aleatorios

In [26]:
# Fijar semilla para reproducibilidad
np.random.seed(42)

# Aleatorios uniformes [0, 1)
rand_uniform = np.random.random((3, 3)) # matriz de 3x3 con valores aleatorios uniformes
print("Random uniforme (3x3):")
print(rand_uniform)

# Enteros aleatorios
rand_int = np.random.randint(0, 100, size=(3, 4)) # matriz de 3x4 con enteros aleatorios entre 0 y 99
print("\nRandom enteros [0, 100) (3x4):")
print(rand_int)

# Distribución normal (media=0, std=1)
rand_normal = np.random.randn(3, 3) # matriz de 3x3 con valores aleatorios de una distribución normal
print("\nRandom normal (3x3):")
print(rand_normal)

Random uniforme (3x3):
[[0.37454012 0.95071431 0.73199394]
 [0.59865848 0.15601864 0.15599452]
 [0.05808361 0.86617615 0.60111501]]

Random enteros [0, 100) (3x4):
[[23  2 21 52]
 [ 1 87 29 37]
 [ 1 63 59 20]]

Random normal (3x3):
[[ 1.46237812  1.53871497 -2.43910582]
 [ 0.60344123 -0.25104397 -0.16386712]
 [-1.47632969  1.48698096 -0.02445518]]


In [18]:
np.random.seed(91)
lista_aleatorio = np.random.randint(1, 1001, size=100)
print("\nLista de 100 enteros aleatorios entre 1 y 1000:")
print(lista_aleatorio)


Lista de 100 enteros aleatorios entre 1 y 1000:
[179 687 875 211 862 117 798 495 805 870 643 157 683 502 783  29 914 146
 985 507 578 531 661 809 343 478 882 619 149 623 658 510 300 199 692 655
 837 318 803 250 632  79 399 848 882   9  84 331 111  54 269 240 803 323
 452 827 684 725 887 932 216 944 978 523 727 641 762 771  40 797 606 545
 119 181 755 296 495   1 976  64 480 400 868 611 329  58 365 319 175 237
 807 856 139 201 263 863 305 892 115 884]


---

## 3. Atributos y Propiedades de Arrays

Todo array NumPy tiene atributos fundamentales que se deben conocer.

In [30]:
# Crear array de ejemplo
arr = np.random.randint(0, 100, size=(3, 4, 5)) # matriz de 3x4x5 con enteros aleatorios entre 0 y 99

print("Array original:")
print(arr)
print("\nAtributos:")
print(f"ndim (dimensiones): {arr.ndim}")
print(f"shape (forma): {arr.shape}")
print(f"size (total elementos): {arr.size}")
print(f"dtype (tipo de dato): {arr.dtype}")
print(f"itemsize (bytes por elemento): {arr.itemsize}")
print(f"nbytes (bytes totales): {arr.nbytes}")

Array original:
[[[91 53 86 95 96]
  [ 0 18  1 52 43]
  [89 31 69 31 67]
  [54 74 55 16 37]]

 [[23 68 97 69 85]
  [10 15 96 72 58]
  [69 79 92  2 19]
  [58 35 18 89 66]]

 [[18 19 95 70 51]
  [32 39 38 81  0]
  [10 91 56 88 49]
  [22 30 93 41 98]]]

Atributos:
ndim (dimensiones): 3
shape (forma): (3, 4, 5)
size (total elementos): 60
dtype (tipo de dato): int64
itemsize (bytes por elemento): 8
nbytes (bytes totales): 480


In [24]:
class ArrayInfo:
    def __init__ (self, array: np.ndarray):
        self.array = array
        self.ndim = array.ndim
        self.shape = array.shape
        self.size = array.size
        self.dtype = array.dtype
        self.itemsize = array.itemsize
        self.nbytes = array.nbytes
    
    def printInfo(self):
        print("Array original:")
        print(self.array)
        print("\nAtributos:")
        print(f"ndim (dimensiones): {self.ndim}")
        print(f"shape (forma): {self.shape}")
        print(f"size (total elementos): {self.size}")
        print(f"dtype (tipo de dato): {self.dtype}")
        print(f"itemsize (bytes por elemento): {self.itemsize}")
        print(f"nbytes (bytes totales): {self.nbytes}")

arr_1 = np.random.randint(0, 101, size=(3, 4))
arr_3 = np.random.randint(0, 101, size=(2, 3, 4))
info_1 = ArrayInfo(arr_1)
info_2 = ArrayInfo(arr_3)
info_1.printInfo()
print("\n")
info_2.printInfo()

Array original:
[[ 1  7  0 58]
 [91 21 52 19]
 [ 8 53 76 26]]

Atributos:
ndim (dimensiones): 2
shape (forma): (3, 4)
size (total elementos): 12
dtype (tipo de dato): int64
itemsize (bytes por elemento): 8
nbytes (bytes totales): 96


Array original:
[[[ 59  86  86  64]
  [ 16  71  71  82]
  [  6  51  40  40]]

 [[ 48  86  95  78]
  [ 16 100  72  24]
  [ 16  67  74  47]]]

Atributos:
ndim (dimensiones): 3
shape (forma): (2, 3, 4)
size (total elementos): 24
dtype (tipo de dato): int64
itemsize (bytes por elemento): 8
nbytes (bytes totales): 192


### 3.1 Tipos de datos (dtype)

In [32]:
# Especificar tipo de dato
arr_int32 = np.array([1, 2, 3], dtype=np.int32)
arr_float64 = np.array([1, 2, 3], dtype=np.float64)
arr_complex = np.array([1, 2, 3], dtype=np.complex128)

print("int32:", arr_int32.dtype)
print("float64:", arr_float64.dtype)
print("complex128:", arr_complex.dtype)

# Conversión de tipos
arr_float = arr_int32.astype(np.float32)
print("\nConvertido a float32:", arr_float.dtype)

int32: int32
float64: float64
complex128: complex128

Convertido a float32: float32


In [26]:
array = np.array([2.5, 3.5, 4.5])
print("Array original:", array)
print("Tipo de dato original:", array.dtype)

array_int = array.astype(np.int64)
print("Array convertido:", array_int)
print("Tipo de dato convertido:", array_int.dtype)

Array original: [2.5 3.5 4.5]
Tipo de dato original: float64
Array convertido: [2 3 4]
Tipo de dato convertido: int64


---

## 4. Indexación y Slicing

### 4.1 Indexación básica (1D)

In [34]:
arr = np.array([10, 20, 30, 40, 50, 60, 70])

print("Array:", arr)
print("arr[0]:", arr[0])
print("arr[-1]:", arr[-1])
print("arr[2:5]:", arr[2:5])
print("arr[:3]:", arr[:3])
print("arr[::2]:", arr[::2])  # cada 2 elementos
print("arr[::-1]:", arr[::-1])  # invertir

Array: [10 20 30 40 50 60 70]
arr[0]: 10
arr[-1]: 70
arr[2:5]: [30 40 50]
arr[:3]: [10 20 30]
arr[::2]: [10 30 50 70]
arr[::-1]: [70 60 50 40 30 20 10]


In [29]:
arr = np.array([1, 2, 3, 4, 5])
arr_reverse = arr[::-1]
arr_reverse_reverse = arr_reverse[::-1]
print("Array original:", arr)
print("Array invertido:", arr_reverse)
print("Array invertido de nuevo:", arr_reverse_reverse)
print("¿El array invertido de nuevo es igual al original? (np.array_equal): ", np.array_equal(arr, arr_reverse_reverse))

Array original: [1 2 3 4 5]
Array invertido: [5 4 3 2 1]
Array invertido de nuevo: [1 2 3 4 5]
¿El array invertido de nuevo es igual al original? (np.array_equal):  True


### 4.2 Indexación multidimensional

In [30]:
# Matriz 2D
matriz = np.array([[1, 2, 3, 4],
                   [5, 6, 7, 8],
                   [9, 10, 11, 12]])

print("Matriz:")
print(matriz)
print("\nmatriz[1, 2] (fila 1, col 2):", matriz[1, 2])
print("\nPrimera fila:")
print(matriz[0, :])
print("\nSegunda columna:")
print(matriz[:, 1])
print("\nSubmatriz (primeras 2 filas, primeras 3 cols):")
print(matriz[:2, :3])

Matriz:
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]

matriz[1, 2] (fila 1, col 2): 7

Primera fila:
[1 2 3 4]

Segunda columna:
[ 2  6 10]

Submatriz (primeras 2 filas, primeras 3 cols):
[[1 2 3]
 [5 6 7]]


### 4.3 Indexación booleana

In [31]:
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

# Crear máscara booleana
mascara = arr > 5
print("Array:", arr)
print("Máscara (arr > 5):", mascara)
print("Valores > 5:", arr[mascara])

# Directamente
print("\nPares:", arr[arr % 2 == 0])
print("Entre 3 y 7:", arr[(arr >= 3) & (arr <= 7)])

Array: [ 1  2  3  4  5  6  7  8  9 10]
Máscara (arr > 5): [False False False False False  True  True  True  True  True]
Valores > 5: [ 6  7  8  9 10]

Pares: [ 2  4  6  8 10]
Entre 3 y 7: [3 4 5 6 7]


In [36]:
arr = np.array([0, 2, 4, 6, 8, 10])

mascara = (arr >= 3) & (arr % 2 == 0)
valores_mascara = arr[mascara]
print("\nMáscara (arr >= 3 y par):", mascara)
print("Valores que cumplen máscara 2:", valores_mascara)

arr = np.random.randint(0, 1000, size=100)
filtro = (arr % 3 == 0) & (arr % 7 == 0)
valores_filtrados = arr[filtro]
print("\nValores entre 0 y 999 que son múltiplos de 3 y 7:")
print(valores_filtrados)


Máscara (arr >= 3 y par): [False False  True  True  True  True]
Valores que cumplen máscara 2: [ 4  6  8 10]

Valores entre 0 y 999 que son múltiplos de 3 y 7:
[399 273 588  63 105]


### 4.4 Indexación fancy (con arrays)

In [38]:
arr = np.array([10, 20, 30, 40, 50, 60, 70])

# Seleccionar índices específicos
indices = [0, 2, 5]
print("Array:", arr)
print("Elementos en índices [0, 2, 5]:", arr[indices])

# En 2D
matriz = np.arange(12).reshape(3, 4)
print("\nMatriz:")
print(matriz)
print("\nFilas [0, 2]:")
print(matriz[[0, 2]])

Array: [10 20 30 40 50 60 70]
Elementos en índices [0, 2, 5]: [10 30 60]

Matriz:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]

Filas [0, 2]:
[[ 0  1  2  3]
 [ 8  9 10 11]]


In [11]:
array = np.array([1, 3, 5, 0, 1, 54, 21])

extract = array[[0, 5, 1]]
print("Array original:", array)
print("Elementos extraídos (índices 0, 5, 1):", extract)

Array original: [ 1  3  5  0  1 54 21]
Elementos extraídos (índices 0, 5, 1): [ 1 54  3]


---

## 5. Operaciones Matemáticas y Broadcasting

### 5.1 Operaciones elemento a elemento

In [39]:
a = np.array([1, 2, 3, 4])
b = np.array([10, 20, 30, 40])

print("a:", a)
print("b:", b)
print("\na + b:", a + b)
print("a - b:", a - b)
print("a * b:", a * b)
print("a / b:", a / b)
print("a ** 2:", a ** 2)
print("a % 2:", a % 2)

a: [1 2 3 4]
b: [10 20 30 40]

a + b: [11 22 33 44]
a - b: [ -9 -18 -27 -36]
a * b: [ 10  40  90 160]
a / b: [0.1 0.1 0.1 0.1]
a ** 2: [ 1  4  9 16]
a % 2: [1 0 1 0]


In [12]:
class ArrayOperations:
    def __init__(self, array1: np.ndarray, array2: np.ndarray):
        self.array1 = array1
        self.array2 = array2
    
    def add(self):
        return self.array1 + self.array2
    
    def subtract(self):
        return self.array1 - self.array2
    
    def multiply(self):
        return self.array1 * self.array2
    
    def divide(self):
        return self.array1 / self.array2
    
    def power(self, exponent):
        return self.array1 ** exponent
    
    def modulus(self, divisor):
        return self.array1 % divisor
    
arr_a = np.array([1, 2, 3, 4])
arr_b = np.array([10, 20, 30, 40])

operations = ArrayOperations(arr_a, arr_b)
print("a + b:", operations.add())
print("a - b:", operations.subtract())
print("a * b:", operations.multiply())
print("a / b:", operations.divide())
print("a ** 2:", operations.power(2))
print("a % 2:", operations.modulus(2))

a + b: [11 22 33 44]
a - b: [ -9 -18 -27 -36]
a * b: [ 10  40  90 160]
a / b: [0.1 0.1 0.1 0.1]
a ** 2: [ 1  4  9 16]
a % 2: [1 0 1 0]


### 5.2 Broadcasting

Broadcasting permite realizar operaciones entre arrays de diferentes formas.

In [3]:
# Escalar con array
arr = np.array([1, 2, 3, 4])
print("arr:", arr)
print("arr * 10:", arr * 10)

# Vector con matriz
matriz = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]])
vector = np.array([10, 20, 30])

print("\nMatriz:")
print(matriz)
print("\nVector:", vector)
print("\nMatriz + vector (broadcast):")
print(matriz + vector)

# Broadcasting en diferentes ejes
col_vector = np.array([[10], [20], [30]])
print("\nVector columna:")
print(col_vector)
print("\nMatriz * vector columna:")
print(matriz * col_vector)

arr: [1 2 3 4]
arr * 10: [10 20 30 40]

Matriz:
[[1 2 3]
 [4 5 6]
 [7 8 9]]

Vector: [10 20 30]

Matriz + vector (broadcast):
[[11 22 33]
 [14 25 36]
 [17 28 39]]

Vector columna:
[[10]
 [20]
 [30]]

Matriz * vector columna:
[[ 10  20  30]
 [ 80 100 120]
 [210 240 270]]


### 5.3 Operaciones matriciales

In [13]:
A = np.array([[1, 2, 5],
              [3, 4, 6],
              [7, 8, 9]])

B = np.array([[5, 6, 10],
              [7, 8, 11],
              [12, 13, 14]])

print("A:")
print(A)
print("\nB:")
print(B)

# Producto elemento a elemento
print("\nA * B (elemento a elemento):")
print(A * B)

# Producto matricial (dot product)
print("\nA @ B (producto matricial):")
print(A @ B)

# También con np.dot
print("\nnp.dot(A, B):")
print(np.dot(A, B))

# Transpuesta
print("\nA.T (transpuesta):")
print(A.T)

A:
[[1 2 5]
 [3 4 6]
 [7 8 9]]

B:
[[ 5  6 10]
 [ 7  8 11]
 [12 13 14]]

A * B (elemento a elemento):
[[  5  12  50]
 [ 21  32  66]
 [ 84 104 126]]

A @ B (producto matricial):
[[ 79  87 102]
 [115 128 158]
 [199 223 284]]

np.dot(A, B):
[[ 79  87 102]
 [115 128 158]
 [199 223 284]]

A.T (transpuesta):
[[1 3 7]
 [2 4 8]
 [5 6 9]]


---

## 6. Funciones Universales (ufuncs)

Funciones optimizadas que operan elemento a elemento.

In [14]:
arr = np.array([0, 1, 2, 3, 4, 5])

print("arr:", arr)
print("\nFunciones matemáticas:")
print("np.sqrt(arr):", np.sqrt(arr))
print("np.exp(arr):", np.exp(arr))
print("np.log(arr + 1):", np.log(arr + 1))  # +1 para evitar log(0)
print("np.sin(arr):", np.sin(arr))
print("np.cos(arr):", np.cos(arr))

# Funciones de redondeo
decimales = np.array([1.234, 5.678, 9.012])
print("\ndecimals:", decimales)
print("np.round(decimales, 1):", np.round(decimales, 1))
print("np.floor(decimales):", np.floor(decimales))
print("np.ceil(decimales):", np.ceil(decimales))

# Funciones estadísticas
data = np.array([10, 20, 30, 40, 50])
print("\nData:", data)
print("np.mean(data):", np.mean(data))
print("np.median(data):", np.median(data))
print("np.std(data):", np.std(data))
print("np.var(data):", np.var(data))

arr: [0 1 2 3 4 5]

Funciones matemáticas:
np.sqrt(arr): [0.         1.         1.41421356 1.73205081 2.         2.23606798]
np.exp(arr): [  1.           2.71828183   7.3890561   20.08553692  54.59815003
 148.4131591 ]
np.log(arr + 1): [0.         0.69314718 1.09861229 1.38629436 1.60943791 1.79175947]
np.sin(arr): [ 0.          0.84147098  0.90929743  0.14112001 -0.7568025  -0.95892427]
np.cos(arr): [ 1.          0.54030231 -0.41614684 -0.9899925  -0.65364362  0.28366219]

decimals: [1.234 5.678 9.012]
np.round(decimales, 1): [1.2 5.7 9. ]
np.floor(decimales): [1. 5. 9.]
np.ceil(decimales): [ 2.  6. 10.]

Data: [10 20 30 40 50]
np.mean(data): 30.0
np.median(data): 30.0
np.std(data): 14.142135623730951
np.var(data): 200.0


### 6.1 Funciones de agregación

In [15]:
arr = np.array([[1, 2, 3, 4],
                [5, 6, 7, 8],
                [9, 10, 11, 12]])

print("Array:")
print(arr)
print("\nAgregaciones globales:")
print("sum:", arr.sum())
print("min:", arr.min())
print("max:", arr.max())
print("mean:", arr.mean())
print("std:", arr.std())

print("\nPor eje:")
print("sum por columnas (axis=0):", arr.sum(axis=0))
print("sum por filas (axis=1):", arr.sum(axis=1))
print("mean por columnas:", arr.mean(axis=0))
print("max por filas:", arr.max(axis=1))

Array:
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]

Agregaciones globales:
sum: 78
min: 1
max: 12
mean: 6.5
std: 3.452052529534663

Por eje:
sum por columnas (axis=0): [15 18 21 24]
sum por filas (axis=1): [10 26 42]
mean por columnas: [5. 6. 7. 8.]
max por filas: [ 4  8 12]


---

## 7. Manipulación de Forma y Dimensiones

### 7.1 Reshape

In [16]:
arr = np.arange(12)
print("Array original (12,):", arr)

# Reshape a 2D
arr_2d = arr.reshape(3, 4)
print("\nReshape a (3, 4):")
print(arr_2d)

# Reshape a 3D
arr_3d = arr.reshape(2, 2, 3)
print("\nReshape a (2, 2, 3):")
print(arr_3d)

# Usar -1 para inferir dimensión
arr_auto = arr.reshape(4, -1)  # -1 se calcula automáticamente
print("\nReshape a (4, -1):")
print(arr_auto)

Array original (12,): [ 0  1  2  3  4  5  6  7  8  9 10 11]

Reshape a (3, 4):
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]

Reshape a (2, 2, 3):
[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]]

Reshape a (4, -1):
[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]


In [ ]:
array = np.array([1, 2, 3, 4, 5, 4])
variante_1 = array.reshape(6, 1)
variante_2 = array.reshape(2, 3)
variante_3 = array.reshape(3, 2)
print("Array original:", array)
print("\nVariante 1 (6, 1):")
print(variante_1)
print("\nVariante 2 (2, 3):")
print(variante_2)
print("\nVariante 3 (3, 2):")
print(variante_3)

Array original: [1 2 3 4 5 4]

Variante 1 (6, 1):
[[1]
 [2]
 [3]
 [4]
 [5]
 [4]]

Variante 2 (2, 3):
[[1 2 3]
 [4 5 4]]

Variante 3 (3, 2):
[[1 2]
 [3 4]
 [5 4]]


In [63]:
import random
array_size = random.randint(0, 100)

array = np.random.randint(0, 100, size=array_size)

key_value = 4

if array.size % key_value == 0:
    variante = array.reshape(key_value, -1)
    print("Array original:", array)
    print("Tamaño del array:", array.size)
    print("\nVariante reshape(%d, -1):" % key_value)
    print(variante)
else:
    print("El tamaño del array no es múltiplo de %d, no se puede reshape a (%d, -1)." % (key_value, key_value))

Array original: [36 71 79  5 87 55 71 82 97 31 42 60  7  0 42  5  0  7 13 95 80  0 96 93
 27 45 19 80  1 38 78 36 73 77 50 74 55 54 14 93  2 49 96 20 20 62 87 70
 96 74 61 86 44 62 59 79 52 10  8 41  9 87 17  7 15 76 58 21 69 54 57 20
 93 48 51 75  1 28 55 27 99 73 81 22]
Tamaño del array: 84

Variante reshape(4, -1):
[[36 71 79  5 87 55 71 82 97 31 42 60  7  0 42  5  0  7 13 95 80]
 [ 0 96 93 27 45 19 80  1 38 78 36 73 77 50 74 55 54 14 93  2 49]
 [96 20 20 62 87 70 96 74 61 86 44 62 59 79 52 10  8 41  9 87 17]
 [ 7 15 76 58 21 69 54 57 20 93 48 51 75  1 28 55 27 99 73 81 22]]


### 7.2 Flatten y Ravel

In [68]:
matriz = np.array([[1, 2, 3, 4, 5, 6]])

print("Matriz original:")
print(matriz)

# Flatten (crea copia)
flat = matriz.flatten()
print("\nFlatten:", flat)

# Flatten crea una copia, modificarlo no afecta a la matriz original
flat[0] = 88
print("\nDespués de modificar flatten:")
print("Flatten modificado:", flat)
print("Matriz original después de modificar flatten:", matriz)

# Ravel (vista, más eficiente)
rav = matriz.ravel()
print("\nRavel:", rav)

# Modificar ravel afecta a la matriz original
rav[0] = 99
print("\nDespués de modificar ravel:")
print("Ravel modificado:", rav)
print("Matriz original después de modificar ravel:", matriz)

Matriz original:
[[1 2 3 4 5 6]]

Flatten: [1 2 3 4 5 6]

Después de modificar flatten:
Flatten modificado: [88  2  3  4  5  6]
Matriz original después de modificar flatten: [[1 2 3 4 5 6]]

Ravel: [1 2 3 4 5 6]

Después de modificar ravel:
Ravel modificado: [99  2  3  4  5  6]
Matriz original después de modificar ravel: [[99  2  3  4  5  6]]


### 7.3 Concatenar y Apilar

In [72]:
a = np.array([[1, 2],
              [3, 4]])
b = np.array([[5, 6],
              [7, 8]])

print("a:")
print(a)
print("\nb:")
print(b)

# Concatenar verticalmente
v_concat = np.vstack((a, b))
print("\nvstack (vertical):")
print(v_concat)

# Concatenar horizontalmente
h_concat = np.hstack((a, b))
print("\nhstack (horizontal):")
print(h_concat)

a:
[[1 2]
 [3 4]]

b:
[[5 6]
 [7 8]]

vstack (vertical):
[[1 2]
 [3 4]
 [5 6]
 [7 8]]

hstack (horizontal):
[[1 2 5 6]
 [3 4 7 8]]


In [74]:
a = np.array([[4, 5],
              [1, 7]])
b = np.array([[6, 9],
              [3, 1]])

# con axis, axis=0 concatena verticalmente, axis=1 concatena horizontalmente
concat_axis0 = np.concatenate((a, b), axis=0)
print("\nconcatenate axis=0:")
print(concat_axis0)

concat_axis1 = np.concatenate((a, b), axis=1)
print("\nconcatenate axis=1:")
print(concat_axis1)


concatenate axis=0:
[[4 5]
 [1 7]
 [6 9]
 [3 1]]

concatenate axis=1:
[[4 5 6 9]
 [1 7 3 1]]


### 7.4 Split

In [ ]:
arr = np.arange(16).reshape(4, 4)
print("Array original:")
print(arr)

# Split vertical (por filas)
v_split = np.vsplit(arr, 2) # divide en 2 partes verticalmente (2 filas cada una)
print("\nvsplit en 2:")
print("Primera mitad:")
print(v_split[0])
print("Segunda mitad:")
print(v_split[1])

# Split horizontal (por columnas)
h_split = np.hsplit(arr, 2)
print("\nhsplit en 2:")
print("Primera mitad:")
print(h_split[0])

Array original:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]

vsplit en 2:
Primera mitad:
[[0 1 2 3]
 [4 5 6 7]]
Segunda mitad:
[[ 8  9 10 11]
 [12 13 14 15]]

hsplit en 2:
Primera mitad:
[[ 0  1]
 [ 4  5]
 [ 8  9]
 [12 13]]


---

## 8. Operaciones Estadísticas

NumPy proporciona funciones estadísticas completas.

In [83]:
# Generar datos de ejemplo
np.random.seed(42)
datos = np.random.randn(1000)  # 1000 valores de distribución normal

def head_numpy(arr, n=5):
    return arr[:n]


print("Datos de ejemplo:")
print(head_numpy(datos))

print("Estadísticas descriptivas:")
print(f"Media: {np.mean(datos):.4f}")
print(f"Mediana: {np.median(datos):.4f}")
print(f"Desviación estándar: {np.std(datos):.4f}")
print(f"Varianza: {np.var(datos):.4f}")
print(f"Mínimo: {np.min(datos):.4f}")
print(f"Máximo: {np.max(datos):.4f}")

# Percentiles
print("\nPercentiles:")
print(f"25%: {np.percentile(datos, 25):.4f}")
print(f"50%: {np.percentile(datos, 50):.4f}")
print(f"75%: {np.percentile(datos, 75):.4f}")

# Correlación y covarianza
x = np.random.randn(100)
y = 2 * x + np.random.randn(100) * 0.5  # y correlacionado con x

print("\nCorrelación:")
correlacion = np.corrcoef(x, y)
print(correlacion)

Datos de ejemplo:
[ 0.49671415 -0.1382643   0.64768854  1.52302986 -0.23415337]
Estadísticas descriptivas:
Media: 0.0193
Mediana: 0.0253
Desviación estándar: 0.9787
Varianza: 0.9579
Mínimo: -3.2413
Máximo: 3.8527

Percentiles:
25%: -0.6476
50%: 0.0253
75%: 0.6479

Correlación:
[[1.        0.9571594]
 [0.9571594 1.       ]]


---

## 9. Álgebra Lineal

NumPy incluye funciones de álgebra lineal en `np.linalg`.

In [85]:
# Matriz cuadrada
A = np.array([[1, 2],
              [3, 4]])

print("Matriz A:")
print(A)

# Determinante
det = np.linalg.det(A)
print(f"\nDeterminante: {det}")

# Inversa
A_inv = np.linalg.inv(A)
print("\nInversa:")
print(A_inv)

# Verificar A @ A_inv = I
print("\nA @ A_inv (debe ser identidad):")
print(A @ A_inv)

# Eigenvalores y eigenvectores
eigenvalues, eigenvectors = np.linalg.eig(A)
print("\nEigenvalores:", eigenvalues)
print("\nEigenvectores:")
print(eigenvectors)

Matriz A:
[[1 2]
 [3 4]]

Determinante: -2.0000000000000004

Inversa:
[[-2.   1. ]
 [ 1.5 -0.5]]

A @ A_inv (debe ser identidad):
[[1.0000000e+00 0.0000000e+00]
 [8.8817842e-16 1.0000000e+00]]

Eigenvalores: [-0.37228132  5.37228132]

Eigenvectores:
[[-0.82456484 -0.41597356]
 [ 0.56576746 -0.90937671]]


### 9.1 Resolver sistemas de ecuaciones lineales

In [86]:
# Sistema: 3x + 2y = 7
#          x + 4y = 10

# En forma matricial: A @ x = b
A = np.array([[3, 2],
              [1, 4]])
b = np.array([7, 10])

# Resolver
x = np.linalg.solve(A, b)
print("Solución:")
print(f"x = {x[0]:.2f}")
print(f"y = {x[1]:.2f}")

# Verificar
print("\nVerificación A @ x:")
print(A @ x)

Solución:
x = 0.80
y = 2.30

Verificación A @ x:
[ 7. 10.]


In [89]:
class EcuacionLinealSolver:
    def __init__(self, A: np.ndarray, b: np.ndarray):
        self.A = A
        self.b = b
    
    def solve(self):
        return np.linalg.solve(self.A, self.b)
    
    def verify(self, solution):
        return np.allclose(self.A @ solution, self.b)

A = np.array([[-1, 1],
              [-2, 4]])
b = np.array([1, -8])
solver = EcuacionLinealSolver(A, b)
solution = solver.solve()
print("Solución:")
print(f"x = {solution[0]:.2f}")
print(f"y = {solution[1]:.2f}")
print("\nVerificación A @ solution:")
print(solver.verify(solution))

Solución:
x = -6.00
y = -5.00

Verificación A @ solution:
True


### 9.2 Normas de vectores

In [91]:
v = np.array([3, 4])

# Norma L2 (euclidiana)
norma_l2 = np.linalg.norm(v)
print(f"Vector: {v}")
print(f"Norma L2: {norma_l2}")

# Norma L1
norma_l1 = np.linalg.norm(v, ord=1)
print(f"Norma L1: {norma_l1}")

Vector: [3 4]
Norma L2: 5.0
Norma L1: 7.0


---

## 10. Números Aleatorios

### 10.1 Generación de números aleatorios

In [93]:
# Semilla para reproducibilidad
np.random.seed(42)

# Uniforme [0, 1)
uniform = np.random.random(5)
print("Uniforme [0, 1):", uniform)

# Uniforme [a, b)
uniform_range = np.random.uniform(10, 20, size=5)
print("Uniforme [10, 20):", uniform_range)

# Normal (media, std, tamaño)
normal = np.random.normal(100, 15, size=5) # media=100, std=15, 5 valores
print("Normal (μ=100, σ=15):", normal)

# Enteros
integers = np.random.randint(1, 100, size=10)
print("Enteros [1, 100):", integers)

# Choice: seleccionar de un array
arr = np.array(['A', 'B', 'C', 'D'])
choices = np.random.choice(arr, size=10)
print("Choices:", choices)

# Shuffle: mezclar in-place
deck = np.arange(52)
np.random.shuffle(deck)
print("Baraja mezclada (primeros 10):", deck[:10])

Uniforme [0, 1): [0.37454012 0.95071431 0.73199394 0.59865848 0.15601864]
Uniforme [10, 20): [11.5599452  10.58083612 18.66176146 16.01115012 17.08072578]
Normal (μ=100, σ=15): [ 92.95788421 108.13840065  93.04873461  93.0140537  103.62943407]
Enteros [1, 100): [89 49 91 59 42 92 60 80 15 62]
Choices: ['B' 'C' 'B' 'C' 'D' 'C' 'D' 'D' 'A' 'C']
Baraja mezclada (primeros 10): [ 2  0 26 22 12 46 10 18  4 32]


In [113]:
class Carta:
    def __init__(self, valor, palo):
        self.valor = valor
        self.palo = palo
    
    def __str__(self):
        return f"{self.valor} de {self.palo}"

palos = ['Corazones', 'Diamantes', 'Tréboles', 'Picas']
valores = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']
baraja = [Carta(valor, palo) for palo in palos for valor in valores]
np.random.shuffle(baraja)
print("Cartas mezcladas (primeras 5):")
for carta in baraja[:5]:
    print(carta)

Cartas mezcladas (primeras 5):
9 de Diamantes
3 de Tréboles
J de Diamantes
4 de Corazones
10 de Corazones


### 10.2 Muestreo y permutaciones

In [114]:
# Permutación (no modifica original)
arr = np.arange(10)
permuted = np.random.permutation(arr)
print("Original:", arr)
print("Permutado:", permuted)

# Muestreo sin reemplazo
sample = np.random.choice(arr, size=5, replace=False)
print("Muestra sin reemplazo:", sample)

Original: [0 1 2 3 4 5 6 7 8 9]
Permutado: [6 5 2 8 0 1 7 4 9 3]
Muestra sin reemplazo: [3 6 7 1 0]


---

## 11. Lectura y Escritura de Datos

### 11.1 Archivos de texto (CSV)

In [ ]:
# Crear datos de ejemplo
datos = np.random.randint(0, 100, size=(5, 3))
print("Datos originales:")
print(datos)

# Guardar como CSV
np.savetxt('numpy_data/datos.csv', datos, delimiter=',', fmt='%d') # fmt='%d' para enteros
print("\nDatos guardados en 'numpy_data/datos.csv'")

# Leer CSV
datos_leidos = np.loadtxt('numpy_data/datos.csv', delimiter=',', dtype=int)
print("\nDatos leídos:")
print(datos_leidos)

Datos originales:
[[61 31 29]
 [28 48 44]
 [92 29 15]
 [39 18 17]
 [ 0 77 46]]

Datos guardados en 'numpy_data/datos.csv'

Datos leídos:
[[61 31 29]
 [28 48 44]
 [92 29 15]
 [39 18 17]
 [ 0 77 46]]


### 11.2 Formato binario NumPy (.npy, .npz)

In [126]:
# Guardar un array
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.save('numpy_data/array.npy', arr)
print("Array guardado en 'numpy_data/array.npy'")

# Cargar
arr_loaded = np.load('numpy_data/array.npy')
print("\nArray cargado:")
print(arr_loaded)

# Guardar múltiples arrays
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
np.savez('numpy_data/arrays.npz', primero=a, segundo=b)

# Cargar múltiples
data = np.load('numpy_data/arrays.npz')
print("\nArrays cargados:")
print("primero:", data['primero'])
print("segundo:", data['segundo'])

Array guardado en 'numpy_data/array.npy'

Array cargado:
[[1 2 3]
 [4 5 6]]

Arrays cargados:
primero: [1 2 3]
segundo: [4 5 6]


---

## 12. Técnicas Avanzadas

### 12.1 Vectorización vs Loops

In [129]:
import time

# Crear datos grandes
n = 1000000
a = np.random.rand(n)
b = np.random.rand(n)

# Método 1: Loop (LENTO)
start = time.time()
result_loop = np.zeros(n)
for i in range(n):
    result_loop[i] = a[i] + b[i]
tiempo_loop = time.time() - start

# Método 2: Vectorizado (RÁPIDO)
start = time.time()
result_vec = a + b
tiempo_vec = time.time() - start

print(f"Tiempo con loop: {tiempo_loop:.4f}s")
print(f"Tiempo vectorizado: {tiempo_vec:.4f}s")
print(f"Speedup: {tiempo_loop/tiempo_vec:.1f}x")

Tiempo con loop: 0.1654s
Tiempo vectorizado: 0.0012s
Speedup: 138.6x


### 12.2 Where: Condiciones vectorizadas

In [ ]:
# np.where(condicion, valor_si_true, valor_si_false)
arr = np.array([-2, -1, 0, 1, 2, 3, 4, 5])

# Reemplazar negativos por 0, positivos mantener
result = np.where(arr < 0, 0, arr)
print("Original:", arr)
print("Negativos a 0:", result)

# Ejemplo: categorizar valores
valores = np.random.randint(0, 100, size=10)
categorias = np.where(valores < 30, 'bajo', np.where(valores < 70, 'medio', 'alto'))
print("\nValores:", valores)
print("Categorías:", categorias)

Original: [-2 -1  0  1  2  3  4  5]
Negativos a 0: [0 0 0 1 2 3 4 5]

Valores: [60 13 70 15 80 25 54 90 99  7]
Categorías: ['medio' 'bajo' 'alto' 'bajo' 'alto' 'bajo' 'medio' 'alto' 'alto' 'bajo']


In [133]:
valores = np.random.randint(0, 100, size=100)
categorias = np.where(valores % 2 == 0, 'par', 'impar')
print("\nValores:", valores)
print("Categorías (par/impar):", categorias)


Valores: [40 21  9 17 37 97 24 26 74 61 77  4 22 45 82 66 65 92 95 77 99 49 62  4
 51 88 46 91 68 22 91 40 20 29 70 29 63 25 29 94  7 96 29 95 56 78 84  3
 58 44 14 72  5  6 66 38 26 62 42 76 32 88 22 68 92 76 42 12 22 73 82 78
 70 15 85 87 77 70  0 19 94 25 28  6 59 54 70  1 55  0 82 27 44 50 47 81
 34 44 34 38]
Categorías (par/impar): ['par' 'impar' 'impar' 'impar' 'impar' 'impar' 'par' 'par' 'par' 'impar'
 'impar' 'par' 'par' 'impar' 'par' 'par' 'impar' 'par' 'impar' 'impar'
 'impar' 'impar' 'par' 'par' 'impar' 'par' 'par' 'impar' 'par' 'par'
 'impar' 'par' 'par' 'impar' 'par' 'impar' 'impar' 'impar' 'impar' 'par'
 'impar' 'par' 'impar' 'impar' 'par' 'par' 'par' 'impar' 'par' 'par' 'par'
 'par' 'impar' 'par' 'par' 'par' 'par' 'par' 'par' 'par' 'par' 'par' 'par'
 'par' 'par' 'par' 'par' 'par' 'par' 'impar' 'par' 'par' 'par' 'impar'
 'impar' 'impar' 'impar' 'par' 'par' 'impar' 'par' 'impar' 'par' 'par'
 'impar' 'par' 'par' 'impar' 'impar' 'par' 'par' 'impar' 'par' 'par'
 'impar' 'imp

### 12.3 Unique y Bincount

In [136]:
# Unique: valores únicos
arr = np.array([1, 1, 1, 1, 1, 1,  2, 2, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 6, 6, 6, 7, 8, 8, 9, 9, 9, 9, 9])
unicos, conteos = np.unique(arr, return_counts=True)

print("Array:", arr)
print("Valores únicos:", unicos)
print("Conteos:", conteos)

# Bincount: contar ocurrencias de enteros no negativos
conteos_bin = np.bincount(arr)
print("\nBincount:", conteos_bin)
print("Interpretación: índice=valor, valor=conteo")

Array: [1 1 1 1 1 1 2 2 3 3 3 4 4 4 4 5 5 5 6 6 6 7 8 8 9 9 9 9 9]
Valores únicos: [1 2 3 4 5 6 7 8 9]
Conteos: [6 2 3 4 3 3 1 2 5]

Bincount: [0 6 2 3 4 3 3 1 2 5]
Interpretación: índice=valor, valor=conteo


In [140]:
array = np.random.randint(0, 5, size=100)
unicos, conteos = np.unique(array, return_counts=True)

print("Array:", array)
print("Valores únicos:", unicos)
print("Conteos:", conteos)

Array: [2 3 0 1 2 2 1 3 0 2 1 1 1 4 1 0 1 4 2 3 0 3 4 1 4 4 2 3 1 2 4 3 4 1 3 0 2
 1 1 2 0 2 4 2 3 3 1 0 4 0 0 0 2 2 1 0 2 2 4 4 0 3 0 4 0 1 2 2 3 4 4 4 2 3
 3 4 1 2 2 1 4 4 2 3 1 1 0 3 1 4 0 3 0 1 2 2 4 4 3 0]
Valores únicos: [0 1 2 3 4]
Conteos: [18 21 23 17 21]


### 12.4 Meshgrid: Crear grillas

In [143]:
# Útil para visualización y cálculos en grillas
x = np.array([1, 2, 3])
y = np.array([10, 20])

X, Y = np.meshgrid(x, y)

print("x:", x)
print("y:", y)
print("\nX (grilla):")
print(X)
print("\nY (grilla):")
print(Y)

# Ejemplo: calcular Z = X + Y en toda la grilla
Z = X + Y
print("\nZ = X + Y:")
print(Z)

x: [1 2 3]
y: [10 20]

X (grilla):
[[1 2 3]
 [1 2 3]]

Y (grilla):
[[10 10 10]
 [20 20 20]]

Z = X + Y:
[[11 12 13]
 [21 22 23]]


### 12.5 Aplicar funciones: apply_along_axis

In [144]:
# Aplicar función personalizada a lo largo de un eje
matriz = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]])

# Función personalizada: rango (max - min)
def rango(arr):
    return arr.max() - arr.min()

print("Matriz:")
print(matriz)

# Aplicar por columnas (axis=0)
rangos_col = np.apply_along_axis(rango, axis=0, arr=matriz)
print("\nRango por columnas:", rangos_col)

# Aplicar por filas (axis=1)
rangos_fila = np.apply_along_axis(rango, axis=1, arr=matriz)
print("Rango por filas:", rangos_fila)

Matriz:
[[1 2 3]
 [4 5 6]
 [7 8 9]]

Rango por columnas: [6 6 6]
Rango por filas: [2 2 2]


---

## 13. Ejercicios Prácticos

Resuelve los siguientes ejercicios para consolidar tu conocimiento de NumPy.

### Ejercicio 1: Creación y Manipulación Básica

1. Crea un array 1D con los números del 0 al 99
2. Reshape el array a una matriz 10x10
3. Extrae la diagonal principal
4. Calcula la suma de cada fila
5. Encuentra el valor máximo de cada columna

In [ ]:
# Tu solución aquí


### Ejercicio 2: Operaciones Estadísticas

Genera 1000 números de una distribución normal con media=100 y std=15:

1. Calcula media, mediana, desviación estándar
2. ¿Cuántos valores están entre 85 y 115?
3. Normaliza los datos para que tengan media=0 y std=1
4. Verifica que la normalización funcionó

In [ ]:
# Tu solución aquí


### Ejercicio 3: Indexación Avanzada

Dada la siguiente matriz:

```python
matriz = np.random.randint(1, 100, size=(8, 8))
```

1. Extrae la submatriz del centro (4x4)
2. Reemplaza todos los valores mayores a 50 por 50
3. Crea una máscara booleana de valores pares
4. Calcula la suma de los valores pares y los impares por separado

In [ ]:
# Tu solución aquí


### Ejercicio 4: Broadcasting y Vectorización

1. Crea una matriz de 5x3 con valores aleatorios entre 0 y 1
2. Normaliza cada columna para que tenga media 0 (sin usar loops)
3. Escala cada fila para que la suma sea 1 (sin usar loops)
4. Calcula la distancia euclidiana entre todas las filas

In [ ]:
# Tu solución aquí


### Ejercicio 5: Álgebra Lineal Aplicada

Implementa una regresión lineal simple usando solo NumPy:

Dado y = 3x + 2 + ruido, genera datos sintéticos y recupera los coeficientes.

Pasos:
1. Genera X: 100 valores entre 0 y 10
2. Genera y: y = 3*X + 2 + ruido_normal
3. Calcula los coeficientes usando: β = (X^T X)^(-1) X^T y
4. Compara los coeficientes obtenidos con los reales (3 y 2)

In [ ]:
# Tu solución aquí


### Ejercicio 6: Simulación de Monte Carlo

Estima el valor de π usando simulación de Monte Carlo:

1. Genera N=1,000,000 puntos aleatorios en el cuadrado [0,1] x [0,1]
2. Calcula cuántos caen dentro del círculo de radio 1 centrado en (0,0)
3. Estima π usando: π ≈ 4 * (puntos_dentro / total_puntos)
4. Compara con np.pi

In [ ]:
# Tu solución aquí


### Ejercicio 7: Análisis de Datos Reales

Usa los datos de iris de scikit-learn:

```python
from sklearn.datasets import load_iris
iris = load_iris()
X = iris.data  # shape: (150, 4)
y = iris.target  # shape: (150,)
```

1. Calcula la media y desviación estándar de cada característica
2. Normaliza los datos (z-score)
3. Calcula la matriz de correlación entre las 4 características
4. Para cada clase (0, 1, 2), calcula el centroide (media de cada característica)
5. Encuentra la distancia de cada punto a su centroide de clase

In [ ]:
# Tu solución aquí


### Ejercicio 8: Operaciones Avanzadas

Implementa las siguientes funciones usando NumPy:

1. `softmax(x)`: Función softmax para un array 1D
2. `cosine_similarity(a, b)`: Similitud coseno entre dos vectores
3. `moving_average(arr, window)`: Media móvil con ventana deslizante
4. `standardize_columns(X)`: Estandarizar cada columna de una matriz

Prueba cada función con datos de ejemplo.

In [ ]:
# Tu solución aquí

def softmax(x):
    """Calcula softmax de un array 1D"""
    pass

def cosine_similarity(a, b):
    """Calcula similitud coseno entre dos vectores"""
    pass

def moving_average(arr, window):
    """Calcula media móvil con ventana deslizante"""
    pass

def standardize_columns(X):
    """Estandariza cada columna (media=0, std=1)"""
    pass

---

## Conclusión

Has completado el tutorial de NumPy. Ahora dominas:

- Creación y manipulación de arrays
- Operaciones matemáticas y broadcasting
- Indexación avanzada
- Funciones estadísticas
- Álgebra lineal
- Números aleatorios
- Técnicas de optimización

### Próximos pasos:

1. Practica con datasets reales
2. Combina NumPy con pandas para análisis de datos
3. Usa NumPy como base para Machine Learning con scikit-learn
4. Explora broadcasting avanzado para optimizar código

### Recursos adicionales:

- Documentación oficial: https://numpy.org/doc/
- NumPy for Machine Learning: https://numpy.org/numpy-tutorials/
- Cheat sheet: https://numpy.org/devdocs/user/absolute_beginners.html